sdk/core/core-websockets/src/reliableConnectionClient.ts

```diff
@@ -0,0 +1,435 @@
+      const res = canReconnect && (isRetryable?.(err) ?? true);
+      canReconnect = true;
+      return res;
+    }
+    let openAbortSignal: AbortSignal | undefined = undefined;
+    let openOrClosePromiseResolve: (() => void) | undefined = undefined;
+    let openOrClosePromiseReject: ((reason?: any) => void) | undefined = undefined;
+    function clearResolvers(): void {
+      openOrClosePromiseResolve = undefined;
+      openOrClosePromiseReject = undefined;
+    }
+    function rejectAndReset(reason?: any): void {
+      openOrClosePromiseReject?.(reason);
+      clearResolvers();
+    }
+    function resolveAndReset(): void {
+      openOrClosePromiseResolve?.();
+      clearResolvers();
+    }
+    const reliableConnectionClient: WritableClient = {
+      status: "disconnected",
+      open: async (openOpts: OpenOptions = {}) => {
+        switch (reliableConnectionClient.status) {
+          case "connecting":
+          case "connected": {
+            logger.info(`[${connectionId}] Client is already ${reliableConnectionClient.status}`);
+            return;
+          }
+          case "disconnecting": {
+            const errMsg = `[${connectionId}] Client is already ${reliableConnectionClient.status}`;
+            if (resolveOnUnsuccessful) {
+              logger.warning(errMsg);
+              return;
+            } else {
+              throw createError(errMsg);
+            }
+          }
+          case "disconnected": {
+            break;
+          }
+        }
+        const { abortSignal } = openOpts;
+        openAbortSignal = abortSignal ?? openAbortSignal;
+        async function retryableConnecting(
+          opOptions: AbortableOperationOptions = {},
+        ): Promise<void> {
+          logger.verbose(`[${connectionId}] Opening connection`);
+          client.open({ ...openOpts, ...opOptions });
+          /** installs dummy handlers to activate the logic in the wrappers */
+          reliableConnectionClient.onopen(() => {});
+          reliableConnectionClient.onclose(() => {});
+          reliableConnectionClient.onerror(() => {});
+          reliableConnectionClient.onmessage(() => {});
+          reliableConnectionClient.status = "connecting";
+          logger.verbose(`[${connectionId}] Connecting...`);
+          await createAbortablePromise<void>((resolve, reject) => {
+            openOrClosePromiseResolve = resolve;
+            openOrClosePromiseReject = reject;
+          }, opOptions);
+        }
+        try {
+          await retry(retryableConnecting, `open connection [${connectionId}]`, retryOptions, {
+            isRetryable: isOpenRetryable,
+            abortSignal,
+          });
+        } catch (error) {
+          logger.error(`[${connectionId}] Error opening connection:`, error);
+          rejectAndReset(error);
+          try {
+            await reliableConnectionClient.close({ abortSignal });
+          } catch (closeError) {
+            rejectAndReset(closeError);
```

Comment:
maybe more debug logging throughout here, whenever we change connection state?

Reply:
The very first line of the outer catch is logging. This inner one is not interesting to log context for, logging for `close` errors is already taken care of inside `close`.